tests/array_list_with_map_test.c

```diff
@@ -0,0 +1,186 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/array_list_with_map.h>
+#include <aws/common/string.h>
+
+#include <aws/testing/aws_test_harness.h>
+
+static uint64_t s_hash_string_ptr(const void *item) {
+    const struct aws_string *str = *(const struct aws_string **)item;
+    return aws_hash_string((void *)str);
+}
+
+static bool s_hash_string_ptr_eq(const void *a, const void *b) {
+    const struct aws_string *str_a = *(const struct aws_string **)a;
+    const struct aws_string *str_b = *(const struct aws_string **)b;
+    return aws_string_eq(str_a, str_b);
+}
+
+static int s_array_list_with_map_sanitize_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+
+    struct aws_array_list_with_map list_with_map;
+    ASSERT_SUCCESS(aws_array_list_with_map_init(
+        &list_with_map, allocator, s_hash_string_ptr, aws_ptr_eq, NULL, sizeof(void *), 0));
+    aws_array_list_with_map_clean_up(&list_with_map);
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(array_list_with_map_sanitize_test, s_array_list_with_map_sanitize_fn)
+
+static int s_array_list_with_map_insert_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+    AWS_STATIC_STRING_FROM_LITERAL(foo, "foo");
+    AWS_STATIC_STRING_FROM_LITERAL(bar, "bar");
+    AWS_STATIC_STRING_FROM_LITERAL(foobar, "foobar");
+
+    struct aws_array_list_with_map list_with_map;
+    /* With only 1 initial element. */
+    ASSERT_SUCCESS(aws_array_list_with_map_init(
+        &list_with_map, allocator, s_hash_string_ptr, s_hash_string_ptr_eq, NULL, sizeof(struct aws_string *), 1));
+    ASSERT_SUCCESS(aws_array_list_with_map_insert(&list_with_map, &foobar));
+    ASSERT_SUCCESS(aws_array_list_with_map_insert(&list_with_map, &bar));
+    ASSERT_SUCCESS(aws_array_list_with_map_insert(&list_with_map, &foo));
+
+    /* You cannot have duplicates */
+    ASSERT_FAILS(aws_array_list_with_map_insert(&list_with_map, &foobar));
+
+    /* Check the size */
+    ASSERT_UINT_EQUALS(aws_array_list_with_map_length(&list_with_map), 3);
+
+    aws_array_list_with_map_clean_up(&list_with_map);
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(array_list_with_map_insert_test, s_array_list_with_map_insert_fn)
+
+static int s_array_list_with_map_get_random_fn(struct aws_allocator *allocator, void *ctx) {
```

Comment:
I'd want some fuzz testing to make sure our random function gives a roughly uniform output. Not sure how much work this is though.